x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -124,14 +124,17 @@ public LogicalPlan optimize(LogicalPlan verified) {
         Batch substitutions = new Batch("Substitutions", Limiter.ONCE,
                 new RewritePivot(),
                 new ReplaceRegexMatch(),
-                new ReplaceAggregatesWithLiterals(),
-                new ReplaceCountInLocalRelation()
+                new ReplaceAggregatesWithLiterals()
                 );
 
         Batch refs = new Batch("Replace References", Limiter.ONCE,
                 new ReplaceReferenceAttributeWithSource()
                 );
 
+        Batch pushDownFilters = new Batch("Push down filters",
+                // must happen in a batch before ReplaceAggregationsOnLocalRelations
+                new PushDownAndCombineFilters());
```

Comment:
I now managed to fix the orders of the rules such that `PushDownAndCombineFilters` can remain in the original batch.

My problem is that `ReplaceAggregationsInLocalRelations`, `ConstantFolding` and `PruneFilters` form a series of transformations that must be applied in this sequence before `PushDownAndCombineFilters`. If `PushDownAndCombineFilters` is applied in the middle of this sequence, it pushes down filters it shouldn't. E.g. for the query `SELECT COUNT(*) HAVING COUNT(*) = 0`:

Correct sequence:
```
Having[count(*) = 0]
\ Agg[count(*)]
 \ Local[single]
```
=Fold=>
```
Having[false]
\ Agg[1]
 \ Local[single]
```
=Prune=>
```
Local[empty]
```

Incorrect sequence:
```
Having[count(*) = 0]
\ Agg[count(*)]
 \ Local[single]
```
=Fold=>
```
Having[false]
\ Agg[1]
 \ Local[single]
```
=PushDownFilter=> // pushes down because Having does not contain an aggregation anymore
```
Agg[1]
 \ Having[false]
  \ Local[single]
```
=Prune=>
```
Agg[1]
\ Local[empty]
```

It's like `ReplaceAggregationsInLocalRelations` removes the information that we're dealing with an aggregation and the correct context is only restored with `PruneFilter`.